c++/h5/stl/map.hpp

```diff
@@ -14,38 +14,87 @@
 
 #pragma once
 #include <map>
+#include <type_traits>
 #include "../group.hpp"
 #include "./string.hpp"
 
 namespace h5 {
 
-  template <typename T>
-  struct hdf5_format_impl<std::map<std::string, T>> {
+  template <typename keyT, typename valueT>
+  struct hdf5_format_impl<std::map<keyT, valueT>> {
     static std::string invoke() { return "Dict"; }
   };
 
+
   /**
-   * Map of string and T as a subgroup with key_names
-   */
+   * Map of type keyT for the key and valueT for the value. keyT can be any 
+   * class as long as it is writeable to h5 (an operator "<" is needed to 
+   * be used in a map in the first place).
+  */
+  template <typename keyT, typename valueT>
+  void h5_write(group f, std::string const &name, std::map<keyT, valueT> const &M) {
+    auto gr = f.create_group(name);
+    write_hdf5_format(gr, M);
+    
+    if constexpr (std::is_same_v<keyT, std::string>){
+      for (auto &pvp : M) h5_write(gr, pvp.first, pvp.second);
+    }
+    else {
+      int indx = 0;
+      for (auto &pvp : M) {
+        auto element_gr = gr.create_group( std::to_string(indx) );
+        h5_write(element_gr, "key", pvp.first);
+        h5_write(element_gr, "val", pvp.second);
+        ++indx;
+      }
+    }
+  }
+
+  template <typename keyT, typename valueT>
+  void h5_read(group f, std::string const &name, std::map<keyT, valueT> &M) {
+    auto gr = f.open_group(name);
+    M.clear();
+    
+    if constexpr (std::is_same_v<keyT, std::string>){
+      for (auto const &x : gr.get_all_subgroup_dataset_names()) {
+        valueT value;
```

Comment:
Pull these lines outside of the `if constexpr {..} else {..}`. Consistently use `val` or `value` as a variable name for all changes?